src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -290,25 +290,92 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static ucs_status_t
+uct_cuda_copy_detect_vmm(void *address, CUmemorytype *cuda_mem_type,
+                         unsigned *is_vmm_allocation)
+{
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop = {};
+    CUresult cu_err;
+    CUmemGenericAllocationHandle alloc_handle;
+
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    cu_err = cuMemRetainAllocationHandle(&alloc_handle, (void*)address);
+    if (cu_err == CUDA_SUCCESS) {
+        *is_vmm_allocation = 1;
+        if (*cuda_mem_type == CU_MEMORYTYPE_HOST) {
+            cu_err = cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle);
+            if (cu_err != CUDA_SUCCESS) {
+                cuMemRelease(alloc_handle);
+                return UCS_ERR_INVALID_ADDR;
+            }
+
+            /* if location type is NUMA or NUMA_CURRENT the memory is EGM and can be
+             * potentially accessed by copy engines */
+            if ((prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA) ||
+                (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT)) {
+                *cuda_mem_type = CU_MEMORYTYPE_DEVICE;
+            } else {
```

Comment:
Here we can have `prop.location.type == CU_MEM_LOCATION_TYPE_HOST`. I don't think that we should return error in this case.